dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/apikey/Dhis2ApiTokenFilter.java

```diff
@@ -0,0 +1,248 @@
+        tokenKey = null;
+
+        try
+        {
+            ApiTokenAuthenticationToken authenticationToken = (ApiTokenAuthenticationToken) apiTokenAuthManager
+                .authenticate( new ApiTokenAuthenticationToken( hashedKey ) );
+
+            // Set values unique to each request
+            authenticationToken.setDetails( this.authenticationDetailsSource.buildDetails( request ) );
+
+            validateRequestRules( request, authenticationToken.getToken() );
+
+            authenticationToken.setAuthenticated( true );
+
+            SecurityContext context = SecurityContextHolder.createEmptyContext();
+            context.setAuthentication( authenticationToken );
+            SecurityContextHolder.setContext( context );
+
+            if ( this.logger.isDebugEnabled() )
+            {
+                this.logger.debug( LogMessage.format( "Set SecurityContextHolder to %s", authenticationToken ) );
+            }
+
+            filterChain.doFilter( request, response );
+        }
+        catch ( AuthenticationException failed )
+        {
+            SecurityContextHolder.clearContext();
+
+            this.logger.debug( "Failed to process authentication request", failed );
+            this.authenticationFailureHandler.onAuthenticationFailure( request, response, failed );
+        }
+    }
+
+    private void validateRequestRules( HttpServletRequest request, ApiToken token )
+    {
+        final List<String> errors = new ArrayList<>();
+
+        for ( ApiTokenAttribute attribute : token.getAttributes() )
+        {
+            if ( attribute instanceof IpAllowedList )
+            {
+                validateIp( request, errors, (IpAllowedList) attribute );
+            }
+
+            if ( attribute instanceof RefererAllowedList )
+            {
+                validateReferer( request, errors, (RefererAllowedList) attribute );
+            }
+
+            if ( attribute instanceof MethodAllowedList )
+            {
+                validateMethod( request, errors, (MethodAllowedList) attribute );
+            }
+        }
+
+        if ( !errors.isEmpty() )
+        {
+            throw new ApiTokenConstraintsValidationFailedException( errors.get( 0 ) );
+        }
+    }
+
+    private void validateMethod( HttpServletRequest request, List<String> errors, MethodAllowedList attribute )
+    {
+        if ( !attribute.getAllowedMethods().isEmpty() )
+        {
+            String method = request.getMethod();
+
+            if ( !attribute.getAllowedMethods().contains( method ) )
+            {
+                errors.add( "Failed to authenticate API token, request http method is not allowed." );
+            }
+        }
+    }
+
+    private void validateReferer( HttpServletRequest request, List<String> errors, RefererAllowedList attribute )
+    {
+        if ( !attribute.getAllowedReferrers().isEmpty() )
+        {
+            String referrer = request.getHeader( "referrer" );
```

Comment:
This header is actually (mis)spelt [referer](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer) for historical reasons.

```suggestion
            String referrer = request.getHeader( "referer" );
```

Reply:
Nice catch @mediremi